apps/dashboard/src/components/primitives/field-editor/field-editor.tsx

```diff
@@ -0,0 +1,232 @@
+import { EditorView } from '@uiw/react-codemirror';
+import { Completion, CompletionContext } from '@codemirror/autocomplete';
+import { autocompletion } from '@codemirror/autocomplete';
+
+import { Editor } from '@/components/primitives/editor';
+import { completions } from '@/utils/liquid-autocomplete';
+import { LiquidVariable } from '@/utils/parseStepVariablesToLiquidVariables';
+import { Popover, PopoverTrigger } from '@/components/primitives/popover';
+import { createVariablePlugin } from './variable-plugin';
+import { VariablePopover } from './variable-popover';
+import { variablePillTheme } from './variable-plugin/variable-theme';
+import { useCallback, useMemo, useRef, useState, Dispatch, SetStateAction } from 'react';
+
+type SelectedVariable = {
+  value: string;
+  from: number;
+  to: number;
+} | null;
+
+type CompletionRange = {
+  from: number;
+  to: number;
+} | null;
+
+type FieldEditorProps = {
+  value: string;
+  onChange: (value: string) => void;
+  variables: LiquidVariable[];
+  placeholder?: string;
+  autoFocus?: boolean;
+  size?: 'default' | 'lg';
+  fontFamily?: 'inherit';
+  id?: string;
+};
+
+export function FieldEditor({
+  value,
+  onChange,
+  variables,
+  placeholder,
+  autoFocus,
+  size = 'default',
+  fontFamily = 'inherit',
+  id,
+}: FieldEditorProps) {
+  const viewRef = useRef<EditorView | null>(null);
+  const lastCompletionRef = useRef<CompletionRange>(null);
+
+  const { selectedVariable, setSelectedVariable, handleVariableSelect, isUpdatingRef } = useVariableSelection();
+
+  const handleVariableUpdate = useVariableUpdate(
```

Comment:
I strongly suggest merging `useVariableSelection` and `useVariableSelection` into a `useVariables` hook. The hooks are strongly connected, as the output of the first one is the input of the second. Moreover, they share the same state using common ref objects.